draco/analysis/dayenu.py

```diff
@@ -0,0 +1,746 @@
+    uflag = uflag[:, np.newaxis, :] & uflag[:, :, np.newaxis]
+    uflag = uflag.astype(np.float64)
+
+    ucov = uflag * cov[np.newaxis, :, :]
+
+    pinv = np.linalg.pinv(ucov, hermitian=True) * uflag
+    pinv = pinv[uindex, :, :].reshape(oshp)
+
+    return pinv
+
+
+def lowpass_mmode_filter(ra, m_cut, flag, epsilon=1e-10):
+    """Construct a low-pass m-mode filter.
+
+    The pass band will range from [-m_cut, m_cut].
+
+    Parameters
+    ----------
+    ra : np.ndarray[nra,]
+        Righ ascension in radians.
+    m_cut : float
+        The half width of the pass band.
+    flag : np.ndarray[nfreq, nra]
+        Boolean flag that indicates what right ascensions are valid
+        as a function of frequency.
+    epsilon : float
+        The stop-band rejection of the filter.  Defaults to 1e-10.
+
+    Returns
+    -------
+    pinv : np.ndarray[nfreq, nra, nra]
+        Low-pass m-mode filter for each frequency.
+    """
+    ishp = flag.shape
+    nra = ra.size
+    assert ishp[-1] == nra
+
+    oshp = ishp + (nra,)
+
+    a = np.median(np.abs(np.diff(ra))) * m_cut / np.pi
+    aeps = a * epsilon
+
+    dra = ra[:, np.newaxis] - ra[np.newaxis, :]
+
+    cov = np.eye(nra, dtype=np.float64) / aeps
+    cov += a * (1.0 - 1.0 / aeps) * np.sinc(m_cut * dra / np.pi)
+
+    uflag, uindex = np.unique(flag.reshape(-1, nra), return_inverse=True, axis=0)
+    uflag = uflag[:, np.newaxis, :] & uflag[:, :, np.newaxis]
+    uflag = uflag.astype(np.float64)
+
+    ucov = uflag * cov[np.newaxis, :, :]
+
+    pinv = np.linalg.pinv(ucov, hermitian=True) * uflag
+    pinv = pinv[uindex, :, :].reshape(oshp)
+
+    return pinv
+
+
+def highpass_mmode_filter(ra, m_cut, flag, epsilon=1e-10):
+    """Construct a high-pass m-mode filter.
+
+    The stop band will range from [-m_cut, m_cut].
+
+    Parameters
+    ----------
+    ra : np.ndarray[nra,]
+        Righ ascension in radians.
+    m_cut : float
+        The half width of the stop band.
+    flag : np.ndarray[nfreq, nra]
+        Boolean flag that indicates what right ascensions are valid
+        as a function of frequency.
+    epsilon : float
+        The stop-band rejection of the filter.  Defaults to 1e-10.
+
+    Returns
+    -------
+    pinv : np.ndarray[nfreq, nra, nra]
+        High-pass m-mode filter for each frequency.
+    """
+    ishp = flag.shape
+    nra = ra.size
+    assert ishp[-1] == nra
+
+    oshp = ishp + (nra,)
+
+    dra = ra[:, np.newaxis] - ra[np.newaxis, :]
+
+    cov = np.eye(nra, dtype=np.float64)
+    cov += np.sinc(m_cut * dra / np.pi) / epsilon
+
+    uflag, uindex = np.unique(flag.reshape(-1, nra), return_inverse=True, axis=0)
+    uflag = uflag[:, np.newaxis, :] & uflag[:, :, np.newaxis]
+    uflag = uflag.astype(np.float64)
+
+    ucov = uflag * cov[np.newaxis, :, :]
+
+    pinv = np.linalg.pinv(ucov, hermitian=True) * uflag
+    pinv = pinv[uindex, :, :].reshape(oshp)
+
+    return pinv
+
+
+def instantaneous_m(ha, lat, dec, u, v, w=0.0):
```

Comment:
This seems like probably something we should move to `caput.interferometry` along with the fringe phase code. I wouldn't really delay this PR on that basis though.

Reply:
Yeah that makes sense.  I have been using something like this in several places (constructing the m-space window for the deconvolving ringmap maker and the point source beam calibration).

I can issue a follow up PR that moves this method to `caput`.

Reply:
Sounds good. Although not very high priority!